core/src/main/scala/api/sfs.scala

```diff
@@ -25,15 +30,19 @@ trait Filesystem {
    *  things too eagerly, nor caching data for too long.
    */
   def resolve(path: Path): Key
-  def metadata(key: Key): Metadata
-  def lookup(key: Key): Data
+  def metadata(key: Key): M[Metadata]
+  def lookup(key: Key): M[Data]
```

Comment:
Another thing to consider is that `Path => Metadata` actually has a bunch of hidden evaluation in it if the filesystem is constructed hierarchically. That is, it desugars into `Seq[Name].fold(Root)((parent, name) => parent.md[DirList](name))` or something along those lines. If evaluation must occur to know **which paths** exist in the filesystem, that is unavoidable when such paths are touched. But I think it's low utility to place the entire metadata map behind an effect, because there should always be a metadata map available even if it's empty.

In fact, there can be an attribute key for the fully evaluated metadata map! So this definitely subsumes `M[Metadata]`. Just get your `M[Metadata]` out of the `Metadata`.